File: ios/Detour/ViewModels/RouteViewModel.swift
import SwiftUI
import MapKit
import CoreLocation

@Observable
final class RouteViewModel {
    var originQuery = ""
    var destinationQuery = ""

    var originCoordinate: CLLocationCoordinate2D?
    var destinationCoordinate: CLLocationCoordinate2D?

    var originName: String?
    var destinationName: String?

    var originSuggestions: [MKLocalSearchCompletion] = []
    var destinationSuggestions: [MKLocalSearchCompletion] = []

    var route: MKRoute?
    var poiResults: [POIResult] = []
    var stopResults: [StopResults] = []
    var selectedPOI: POIResult?
    var searchQuery = "coffee"
    var additionalQueries: [String] = []
    var maxDetourMinutes: Double = 15
    var openNowOnly = true
    var selectedCategory: Category? = .coffee
    var travelMode: TravelMode = .drive

    enum TravelMode: String, CaseIterable, Identifiable {
        case drive = "DRIVE"
        case walk = "WALK"
        case bike = "BICYCLE"

        var id: String { rawValue }

        var label: String {
            switch self {
            case .drive: return "Drive"
            case .walk: return "Walk"
            case .bike: return "Bike"
            }
        }

        var icon: String {
            switch self {
            case .drive: return "car.fill"
            case .walk: return "figure.walk"
            case .bike: return "bicycle"
            }
        }

        var mkTransportType: MKDirectionsTransportType? {
            switch self {
            case .drive: return .automobile
            case .walk: return .walking
            case .bike: return nil // MapKit has no bike mode; skip client-side route
            }
        }
    }
    var isLoading = false
    var errorMessage: String?

    var filteredResults: [POIResult] {
        poiResults.filter { poi in
            let withinDetour = poi.detourSeconds <= Int(maxDetourMinutes) * 60
            let openFilter = !openNowOnly || poi.isOpenNow
            let isForward = isAheadOnRoute(poi)
            return withinDetour && openFilter && isForward
        }
    }

    enum Category: String, CaseIterable, Identifiable {
        case coffee = "Coffee"
        case food = "Food"
        case gas = "Gas"
        case grocery = "Grocery"
        case pharmacy = "Pharmacy"
        case evCharging = "EV Charging"

        var id: String { rawValue }

        var icon: String {
            switch self {
            case .coffee: return "cup.and.saucer.fill"
            case .food: return "fork.knife"
            case .gas: return "fuelpump.fill"
            case .grocery: return "cart.fill"
            case .pharmacy: return "cross.case.fill"
            case .evCharging: return "bolt.car.fill"
            }
        }

        var query: String {
            switch self {
            case .coffee: return "coffee"
            case .food: return "restaurant"
            case .gas: return "gas station"
            case .grocery: return "grocery store"
            case .pharmacy: return "pharmacy"
            case .evCharging: return "EV charging station"
            }
        }
    }

    var routeDurationFormatted: String? {
        guard let route else { return nil }
        let minutes = Int(route.expectedTravelTime / 60)
        if minutes >= 60 {
            return "\(minutes / 60)h \(minutes % 60)min"
        }
        return "\(minutes) min"
    }

    var routeDistanceFormatted: String? {
        guard let route else { return nil }
        let km = route.distance / 1000
        if km >= 10 {
            return String(format: "%.0f km", km)
        }
        return String(format: "%.1f km", km)
    }

    var isSearchReady: Bool {
        originCoordinate != nil && destinationCoordinate != nil
    }

    private let originCompleter = SearchCompleterDelegate()
    private let destinationCompleter = SearchCompleterDelegate()

    init() {
        originCompleter.onUpdate = { [weak self] results in
            self?.originSuggestions = results
        }
        destinationCompleter.onUpdate = { [weak self] results in
            self?.destinationSuggestions = results
        }
    }

    func updateOriginQuery(_ query: String) {
        guard query != originQuery else { return }
        originQuery = query
        originCoordinate = nil
        originName = nil
        originCompleter.completer.queryFragment = query
    }

    func updateDestinationQuery(_ query: String) {
        guard query != destinationQuery else { return }
        destinationQuery = query
        destinationCoordinate = nil
        destinationName = nil
        destinationCompleter.completer.queryFragment = query
    }

    func selectOrigin(_ completion: MKLocalSearchCompletion) {
        originQuery = completion.title
        originSuggestions = []
        originCompleter.completer.queryFragment = ""

        Task {
            if let coordinate = await resolveCoordinate(for: completion) {
                await MainActor.run {
                    self.originCoordinate = coordinate
                    self.originName = completion.title
                }
            }
        }
    }

    func selectDestination(_ completion: MKLocalSearchCompletion) {
        destinationQuery = completion.title
        destinationSuggestions = []
        destinationCompleter.completer.queryFragment = ""

        Task {
            if let coordinate = await resolveCoordinate(for: completion) {
                await MainActor.run {
                    self.destinationCoordinate = coordinate
                    self.destinationName = completion.title
                }
            }
        }
    }

    func useCurrentLocation(_ coordinate: CLLocationCoordinate2D) {
        originQuery = "Current Location"
        originCoordinate = coordinate
        originName = "Current Location"
        originSuggestions = []
    }

    private var categoryDebounceTask: Task<Void, Never>?

    func selectCategory(_ category: Category) {
        selectedCategory = category
        searchQuery = category.query
        categoryDebounceTask?.cancel()
        categoryDebounceTask = Task {
            try? await Task.sleep(for: .milliseconds(300))
            guard !Task.isCancelled, isSearchReady else { return }
            await MainActor.run { search() }
        }
    }

    func search() {
        guard let origin = originCoordinate, let destination = destinationCoordinate else { return }
        isLoading = true
        errorMessage = nil
        poiResults = []
        selectedPOI = nil

        Task {
            do {
                if route == nil, let transportType = travelMode.mkTransportType {
                    let dirRequest = MKDirections.Request()
                    dirRequest.source = MKMapItem(placemark: MKPlacemark(coordinate: origin))
                    dirRequest.destination = MKMapItem(placemark: MKPlacemark(coordinate: destination))
                    dirRequest.transportType = transportType

                    let directionsResponse = try await MKDirections(request: dirRequest).calculate()
                    await MainActor.run {
                        self.route = directionsResponse.routes.first
                    }
                }

                let allQueries = additionalQueries.isEmpty ? nil : [searchQuery] + additionalQueries
                let searchResponse = try await APIService.search(
                    origin: (origin.latitude, origin.longitude),
                    destination: (destination.latitude, destination.longitude),
                    query: searchQuery,
                    queries: allQueries,
                    maxDetourMinutes: Int(maxDetourMinutes),
                    openNow: openNowOnly,
                    travelMode: travelMode.rawValue
                )

                await MainActor.run {
                    self.poiResults = searchResponse.results
                    self.stopResults = searchResponse.stops ?? []
                    if searchResponse.results.isEmpty {
                        self.errorMessage = "No places found along this route. Try a different category."
                    }
                    self.isLoading = false
                    AnalyticsService.shared.track("search_executed", properties: [
                        "category": self.selectedCategory?.rawValue ?? "custom",
                        "travelMode": self.travelMode.rawValue,
                        "resultCount": searchResponse.results.count,
                    ])
                    if let o = self.originCoordinate, let d = self.destinationCoordinate {
                        RecentSearchStore.shared.add(
                            originName: self.originName ?? "Origin",
                            origin: o,
                            destinationName: self.destinationName ?? "Destination",
                            destination: d,
                            category: self.searchQuery
                        )
                    }
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = error.localizedDescription
                    self.isLoading = false
                }
            }
        }
    }

    func clear() {
        originQuery = ""
        destinationQuery = ""
        originCoordinate = nil
        destinationCoordinate = nil
        originName = nil
        destinationName = nil
        originSuggestions = []
        destinationSuggestions = []
        route = nil
        poiResults = []
        selectedPOI = nil
        errorMessage = nil
    }

    func swapOriginDestination() {
        swap(&originQuery, &destinationQuery)
        swap(&originCoordinate, &destinationCoordinate)
        swap(&originName, &destinationName)
    }

    func loadSavedRoute(_ saved: SavedRoute) {
        originQuery = saved.originName
        originCoordinate = saved.originCoordinate
        originName = saved.originName
        destinationQuery = saved.destinationName
        destinationCoordinate = saved.destinationCoordinate
        destinationName = saved.destinationName
        originSuggestions = []
        destinationSuggestions = []
        route = nil

        if let cat = Category.allCases.first(where: { $0.query == saved.defaultCategory }) {
            selectedCategory = cat
            searchQuery = cat.query
        }

        search()
    }

    func addStop(_ category: Category) {
        if !additionalQueries.contains(category.query) {
            additionalQueries.append(category.query)
            if isSearchReady { search() }
        }
    }

    func removeStop(at index: Int) {
        guard index < additionalQueries.count else { return }
        additionalQueries.remove(at: index)
        if isSearchReady { search() }
    }

    func clearStops() {
        additionalQueries.removeAll()
        stopResults.removeAll()
    }

    func loadRecentSearch(_ recent: RecentSearch) {
        originQuery = recent.originName
        originCoordinate = recent.originCoordinate
        originName = recent.originName
        destinationQuery = recent.destinationName
        destinationCoordinate = recent.destinationCoordinate
        destinationName = recent.destinationName
        originSuggestions = []
        destinationSuggestions = []
        route = nil

        if let cat = Category.allCases.first(where: { $0.query == recent.category }) {
            selectedCategory = cat
            searchQuery = cat.query
        }

        search()
    }

    func saveCurrentRoute(name: String) -> SavedRoute? {
        guard let origin = originCoordinate,
              let destination = destinationCoordinate,
              let oName = originName,
              let dName = destinationName else { return nil }

        let route = SavedRoute.create(
            name: name,
            origin: origin,
            originName: oName,
            destination: destination,
            destinationName: dName,
            defaultCategory: searchQuery
        )
        SavedRoutesStore.shared.save(route)
        return route
    }

    private func isAheadOnRoute(_ poi: POIResult) -> Bool {
        guard let polyline = route?.polyline else { return true }
        let pointCount = polyline.pointCount
        guard pointCount > 1 else { return true }

        let poiMapPoint = MKMapPoint(poi.coordinate)
        var closestIndex = 0
        var closestDistance = Double.greatestFiniteMagnitude

        for i in 0..<pointCount {
            let routePoint = polyline.points()[i]
            let dist = poiMapPoint.distance(to: routePoint)
            if dist < closestDistance {
                closestDistance = dist
                closestIndex = i
            }
        }

        let progress = Double(closestIndex) / Double(pointCount - 1)
        return progress > 0.05
    }

    private func resolveCoordinate(for completion: MKLocalSearchCompletion) async -> CLLocationCoordinate2D? {
        let request = MKLocalSearch.Request(completion: completion)
        let search = MKLocalSearch(request: request)
        do {
            let response = try await search.start()
            return response.mapItems.first?.placemark.coordinate
        } catch {
            return nil
        }
    }
}

private final class SearchCompleterDelegate: NSObject, MKLocalSearchCompleterDelegate {
    let completer = MKLocalSearchCompleter()
    var onUpdate: (([MKLocalSearchCompletion]) -> Void)?

    override init() {
        super.init()
        completer.delegate = self
        completer.resultTypes = [.address, .pointOfInterest]
    }

    func completerDidUpdateResults(_ completer: MKLocalSearchCompleter) {
        onUpdate?(completer.results)
    }

    func completer(_ completer: MKLocalSearchCompleter, didFailWithError error: Error) {
        onUpdate?([])
    }
}

File: android/app/src/main/java/com/ahmedkhaled/onroute/viewmodel/RouteViewModel.kt
package com.ahmedkhaled.onroute.viewmodel

import android.app.Application
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.ahmedkhaled.onroute.model.*
import com.ahmedkhaled.onroute.model.RecentSearch
import com.ahmedkhaled.onroute.model.RecentSearchStore
import com.ahmedkhaled.onroute.model.SavedRoute
import com.ahmedkhaled.onroute.model.SavedRoutesStore
import com.ahmedkhaled.onroute.service.AnalyticsService
import com.ahmedkhaled.onroute.service.ApiService
import com.ahmedkhaled.onroute.service.DirectionsService
import com.ahmedkhaled.onroute.service.LocationService
import com.ahmedkhaled.onroute.service.PlaceAutocompleteService
import com.ahmedkhaled.onroute.service.PlaceSuggestion
import com.google.android.gms.maps.model.LatLng
import com.google.android.libraries.places.api.Places
import com.google.maps.android.PolyUtil
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class RouteViewModel(application: Application) : AndroidViewModel(application) {

    var originQuery by mutableStateOf("")
        private set
    var destinationQuery by mutableStateOf("")
        private set
    var originLatLng by mutableStateOf<LatLng?>(null)
        private set
    var destinationLatLng by mutableStateOf<LatLng?>(null)
        private set
    var originName by mutableStateOf<String?>(null)
        private set
    var destinationName by mutableStateOf<String?>(null)
        private set

    var originSuggestions by mutableStateOf<List<PlaceSuggestion>>(emptyList())
        private set
    var destinationSuggestions by mutableStateOf<List<PlaceSuggestion>>(emptyList())
        private set

    var travelMode by mutableStateOf(TravelMode.DRIVE)
    var searchQuery by mutableStateOf("coffee")
        private set
    var additionalQueries by mutableStateOf<List<String>>(emptyList())
    var stopResults by mutableStateOf<List<StopResults>>(emptyList())
        private set
    var selectedCategory by mutableStateOf<Category?>(Category.COFFEE)
    var maxDetourMinutes by mutableStateOf(15f)
    var openNowOnly by mutableStateOf(true)

    var routePoints by mutableStateOf<List<LatLng>>(emptyList())
        private set
    var routeDurationFormatted by mutableStateOf<String?>(null)
        private set
    var routeDistanceFormatted by mutableStateOf<String?>(null)
        private set
    var poiResults by mutableStateOf<List<POIResult>>(emptyList())
        private set
    var selectedPOI by mutableStateOf<POIResult?>(null)
    var detourRoutePoints by mutableStateOf<List<LatLng>>(emptyList())
        private set
    var isLoading by mutableStateOf(false)
        private set
    var errorMessage by mutableStateOf<String?>(null)

    val isSearchReady: Boolean
        get() = originLatLng != null && destinationLatLng != null

    val filteredResults: List<POIResult>
        get() = poiResults.filter { poi ->
            val withinDetour = poi.detourSeconds <= (maxDetourMinutes * 60).toInt()
            val openFilter = !openNowOnly || poi.isOpenNow
            val isForward = isAheadOnRoute(poi)
            withinDetour && openFilter && isForward
        }

    private val placesService = PlaceAutocompleteService(Places.createClient(application))
    val locationService = LocationService(application)
    private val apiService = ApiService.create()
    private val directionsService = DirectionsService()
    val savedRoutesStore = SavedRoutesStore(application)
    private val recentSearchStore = RecentSearchStore(application)

    var savedRoutes by mutableStateOf<List<SavedRoute>>(emptyList())
        private set
    var recentSearches by mutableStateOf<List<RecentSearch>>(emptyList())
        private set

    init {
        savedRoutes = savedRoutesStore.load()
        recentSearches = recentSearchStore.load()
    }

    private var originDebounceJob: Job? = null
    private var destinationDebounceJob: Job? = null
    private var originResolveId = 0
    private var destinationResolveId = 0

    fun updateOriginQuery(query: String) {
        if (query == originQuery) return
        originQuery = query
        originLatLng = null
        originName = null
        originDebounceJob?.cancel()
        originDebounceJob = viewModelScope.launch {
            delay(300)
            originSuggestions = placesService.getSuggestions(query)
        }
    }

    fun updateDestinationQuery(query: String) {
        if (query == destinationQuery) return
        destinationQuery = query
        destinationLatLng = null
        destinationName = null
        destinationDebounceJob?.cancel()
        destinationDebounceJob = viewModelScope.launch {
            delay(300)
            destinationSuggestions = placesService.getSuggestions(query)
        }
    }

    fun selectOriginSuggestion(suggestion: PlaceSuggestion) {
        originQuery = suggestion.title
        originSuggestions = emptyList()
        val requestId = ++originResolveId
        viewModelScope.launch {
            placesService.resolvePlace(suggestion.placeId)?.let { (latLng, name) ->
                if (requestId == originResolveId) {
                    originLatLng = latLng
                    originName = name.ifEmpty { suggestion.title }
                }
            }
        }
    }

    fun selectDestinationSuggestion(suggestion: PlaceSuggestion) {
        destinationQuery = suggestion.title
        destinationSuggestions = emptyList()
        val requestId = ++destinationResolveId
        viewModelScope.launch {
            placesService.resolvePlace(suggestion.placeId)?.let { (latLng, name) ->
                if (requestId == destinationResolveId) {
                    destinationLatLng = latLng
                    destinationName = name.ifEmpty { suggestion.title }
                }
            }
        }
    }

    fun useCurrentLocation(latLng: LatLng) {
        originQuery = "Current Location"
        originLatLng = latLng
        originName = "Current Location"
        originSuggestions = emptyList()
    }

    private var categoryDebounceJob: Job? = null

    fun selectCategory(category: Category) {
        selectedCategory = category
        searchQuery = category.query
        categoryDebounceJob?.cancel()
        categoryDebounceJob = viewModelScope.launch {
            delay(300)
            if (isSearchReady) search()
        }
    }

    fun search() {
        val origin = originLatLng ?: return
        val destination = destinationLatLng ?: return
        isLoading = true
        errorMessage = null
        poiResults = emptyList()
        selectedPOI = null

        viewModelScope.launch {
            try {
                val allQueries = if (additionalQueries.isNotEmpty()) listOf(searchQuery) + additionalQueries else null
                val response = apiService.search(
                    SearchRequest(
                        origin = LatLngBody(origin.latitude, origin.longitude),
                        destination = LatLngBody(destination.latitude, destination.longitude),
                        query = searchQuery,
                        queries = allQueries,
                        maxDetourMinutes = maxDetourMinutes.toInt(),
                        openNow = openNowOnly,
                        travelMode = travelMode.apiValue
                    )
                )

                routePoints = PolyUtil.decode(response.route.encodedPolyline)

                val minutes = response.route.durationSeconds / 60
                routeDurationFormatted = if (minutes >= 60) "${minutes / 60}h ${minutes % 60}min" else "$minutes min"
                val km = response.route.distanceMeters / 1000.0
                routeDistanceFormatted = if (km >= 10) "%.0f km".format(km) else "%.1f km".format(km)

                poiResults = response.results
                stopResults = response.stops ?: emptyList()
                if (response.results.isEmpty()) {
                    errorMessage = "No places found along this route. Try a different category."
                }
                isLoading = false
                AnalyticsService.track("search_executed", mapOf(
                    "category" to (selectedCategory?.name ?: "custom"),
                    "travelMode" to travelMode.apiValue,
                    "resultCount" to response.results.size,
                ))
                val o = originLatLng
                val d = destinationLatLng
                if (o != null && d != null) {
                    recentSearchStore.add(
                        originName ?: "Origin", o,
                        destinationName ?: "Destination", d,
                        searchQuery
                    )
                    recentSearches = recentSearchStore.load()
                }
            } catch (e: Exception) {
                errorMessage = e.message ?: "Search failed"
                isLoading = false
            }
        }
    }

    fun swapOriginDestination() {
        val tmpQuery = originQuery
        val tmpLatLng = originLatLng
        val tmpName = originName
        originQuery = destinationQuery
        originLatLng = destinationLatLng
        originName = destinationName
        destinationQuery = tmpQuery
        destinationLatLng = tmpLatLng
        destinationName = tmpName
    }

    fun addStop(category: Category) {
        if (!additionalQueries.contains(category.query)) {
            additionalQueries = additionalQueries + category.query
            if (isSearchReady) search()
        }
    }

    fun removeStop(index: Int) {
        if (index < additionalQueries.size) {
            additionalQueries = additionalQueries.toMutableList().apply { removeAt(index) }
            if (isSearchReady) search()
        }
    }

    fun loadRecentSearch(recent: RecentSearch) {
        originQuery = recent.originName
        originLatLng = recent.originLatLng
        originName = recent.originName
        destinationQuery = recent.destinationName
        destinationLatLng = recent.destinationLatLng
        destinationName = recent.destinationName
        originSuggestions = emptyList()
        destinationSuggestions = emptyList()
        routePoints = emptyList()

        val cat = Category.entries.firstOrNull { it.query == recent.category }
        if (cat != null) {
            selectedCategory = cat
            searchQuery = cat.query
        }

        search()
    }

    fun loadSavedRoute(saved: SavedRoute) {
        originQuery = saved.originName
        originLatLng = saved.originLatLng
        originName = saved.originName
        destinationQuery = saved.destinationName
        destinationLatLng = saved.destinationLatLng
        destinationName = saved.destinationName
        originSuggestions = emptyList()
        destinationSuggestions = emptyList()
        routePoints = emptyList()

        val cat = Category.entries.firstOrNull { it.query == saved.defaultCategory }
        if (cat != null) {
            selectedCategory = cat
            searchQuery = cat.query
        }

        search()
    }

    fun saveCurrentRoute(name: String) {
        val origin = originLatLng ?: return
        val dest = destinationLatLng ?: return
        val route = SavedRoute(
            id = java.util.UUID.randomUUID().toString(),
            name = name,
            originLat = origin.latitude,
            originLng = origin.longitude,
            originName = originName ?: "Origin",
            destinationLat = dest.latitude,
            destinationLng = dest.longitude,
            destinationName = destinationName ?: "Destination",
            defaultCategory = searchQuery,
            createdAt = System.currentTimeMillis()
        )
        savedRoutesStore.save(route)
        savedRoutes = savedRoutesStore.load()
    }

    fun deleteSavedRoute(route: SavedRoute) {
        savedRoutesStore.delete(route)
        savedRoutes = savedRoutesStore.load()
    }

    fun selectPOI(poi: POIResult) {
        selectedPOI = poi
        AnalyticsService.track("poi_selected", mapOf(
            "detourSeconds" to poi.detourSeconds,
            "rating" to poi.rating,
        ))
        val origin = originLatLng ?: return
        val destination = destinationLatLng ?: return

        viewModelScope.launch {
            val detour = directionsService.getDetourRoute(origin, poi.latLng, destination, travelMode.apiValue)
            if (detour != null) {
                detourRoutePoints = detour.points
            }
        }
    }

    fun clearDetour() {
        selectedPOI = null
        detourRoutePoints = emptyList()
    }

    fun clearOrigin() {
        originQuery = ""
        originLatLng = null
        originName = null
        originSuggestions = emptyList()
    }

    fun clearDestination() {
        destinationQuery = ""
        destinationLatLng = null
        destinationName = null
        destinationSuggestions = emptyList()
    }

    private fun isAheadOnRoute(poi: POIResult): Boolean {
        if (routePoints.size < 2) return true
        val poiLatLng = poi.latLng
        var closestIndex = 0
        var closestDist = Double.MAX_VALUE
        for (i in routePoints.indices) {
            val d = distanceBetween(poiLatLng, routePoints[i])
            if (d < closestDist) {
                closestDist = d
                closestIndex = i
            }
        }
        val progress = closestIndex.toDouble() / (routePoints.size - 1)
        return progress > 0.05
    }

    private fun distanceBetween(a: LatLng, b: LatLng): Double {
        val dx = a.latitude - b.latitude
        val dy = a.longitude - b.longitude
        return dx * dx + dy * dy
    }
}
